Mathlib/Data/EENat.lean

```diff
@@ -0,0 +1,161 @@
+      specialize hf n
+      obtain ⟨_, ⟨m, rfl⟩, hm⟩ := hf
+      exact ⟨m, coe_lt_coe.mpr hm⟩
+    | top => exact (lt_irrefl (⊤ : ℕ±∞) ha).elim
+    | bot => exact ⟨Nonempty.some (inferInstance : Nonempty ι), bot_lt_coe _⟩
+
+lemma iSup_coe_lt_top {ι : Sort _} (f : ι → ℕ) : (⨆ x, f x : ℕ±∞) < ⊤ ↔ BddAbove (Set.range f) :=
+  lt_top_iff_ne_top.trans <| (iSup_coe_eq_top f).not.trans not_not
+
+theorem coe_iSup {ι : Sort _} [Nonempty ι] (f : ι → ℕ) (hf : BddAbove (Set.range f)) :
+  ↑(⨆ i, f i) = (⨆ i, f i : ℕ±∞) := by
+  classical
+  apply le_antisymm
+  . rw [le_iSup_iff]
+    refine rec ?_ (fun _ => le_top) ?_
+    . intros n h
+      rw [coe_le_coe, iSup, sSup]
+      change dite _ _ _ ≤_
+      split_ifs with H
+      . apply Nat.find_le
+        rintro _ ⟨m, rfl⟩
+        exact coe_le_coe.mp <| h m
+      . exact zero_le _
+    . intro h
+      have i1 : IsEmpty ι
+      . by_contra r; rw [not_isEmpty_iff] at r; specialize h r.some; rw [le_bot_iff] at h;
+        exact WithBot.coe_ne_bot h
+      exact (not_isEmpty_iff.mpr inferInstance i1).elim
+  . rw [iSup_le_iff]
+    intros i
+    rw [coe_le_coe, iSup]
+    exact le_csSup hf ⟨_, rfl⟩
+
+noncomputable def pred (m : ℕ±∞) := ⨆ x < m, x
+
+lemma pred_le (m : ℕ±∞) : pred m ≤ m := iSup_le fun _ => iSup_le fun hi => le_of_lt hi
+
+lemma pred_bot : pred ⊥ = ⊥ := le_antisymm (pred_le _) bot_le
+lemma pred_top : pred ⊤ = ⊤ := le_antisymm le_top <| le_iSup_iff.mpr fun i hi => by
+  induction i using rec with
+  | coe n =>
+    specialize hi (n + 1)
+    have hi := (le_iSup _ ?_).trans hi
+    . exact (not_lt_of_le hi <| coe_lt_coe.mpr (lt_add_one n)).elim
+    . exact Iff.mp (cmp_eq_lt_iff (n + 1 : ℕ±∞) ⊤) rfl
+  | top => exact le_refl _
+  | bot =>
+    specialize hi 0
+    have hi := (le_iSup _ ?_).trans hi
+    . exact ((not_le_of_lt <| bot_lt_coe 0) hi).elim
+    . exact Iff.mp (cmp_eq_lt_iff 0 ⊤) rfl
+
+lemma pred_zero : pred 0 = ⊥ := Function.swap le_antisymm bot_le <| iSup_le_iff.mpr fun i =>
+  iSup_le_iff.mpr fun hi => by
+    have hi := eq_bot_of_lt_zero _ hi
+    subst hi
+    rfl
+
+lemma pred_succ (n : ℕ) : pred n.succ = n := le_antisymm (iSup_le <| fun i => iSup_le fun hi =>
+  by exact Order.le_of_lt_succ hi) <| le_iSup_of_le n <|
+  le_iSup_of_le (coe_lt_coe.mpr <| Nat.lt.base n) <| le_refl _
+
+lemma pred_le_self (n : ℕ±∞) (hn : n ≤ pred n) : n = ⊥ ∨ n = ⊤ := by
+  have H : ¬ (n = ⊥ ∨ n = ⊤) → ∃ (n' : ℕ), n = n'
+  . intro h
+    rw [not_or] at h
+    induction n using rec with
+    | coe n => exact ⟨n, rfl⟩
+    | top => exact (h.right rfl).elim
+    | bot => exact (h.left rfl).elim
+  contrapose hn
+  specialize H hn
+  rcases H with ⟨n, rfl⟩
+  intro r
+  rcases n with ⟨n⟩
+  . erw [pred_zero] at r
+    exact not_le_of_lt (bot_lt_coe 0) r
+  rw [pred_succ, coe_le_coe] at r
+  exact not_le_of_lt (Nat.lt.step r) r
+
+instance : PredOrder ℕ±∞ where
+  pred := pred
+  pred_le := pred_le
+  min_of_le_pred := fun {m} hm {n} hmn => by
+    have := pred_le_self _ hm
+    sorry
```

Comment:
As has been pointed out, this sorry is false and it's not a PredOrder so we have to use the `pred` function instead of `Order.pred` to transfer between `chainHeight` and `RelSeries.length`